### PR TITLE
feat(sort-structure): #CO-1793 add possibility to sort tickets by structure name

### PR DESCRIPTION
**Pull request by Ozen40**
## Describe your changes
You can now sort tickets by structure name by clicking on "structure du demandeur"
## Checklist tests

- Click on "structure du demandeur" on check if the tickets are correctly sorted by structure name

## Issue ticket number and link
[CO-1793](https://jira.support-ent.fr/browse/CO-1793)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/net/atos/entng/support/services/impl/TicketServiceNeo4jImpl.java [hunk 1]
```diff
@@ -83,4 +87,13 @@ public static Future<JsonObject> listStructureChildren(String structureId) {
         return promise.future();
     }
 
+    public static Future<JsonObject> sortSchoolByName(List<String> structureIds) {
+        Promise<JsonObject> promise = Promise.promise();
+        String query = "MATCH (s:Structure) "
+                + " WHERE s.id IN {structureIds} WITH s ORDER BY s.name RETURN COLLECT(s.id) AS structureIds";
+        Neo4j neo4j = Neo4j.getInstance();
+        JsonObject params = new JsonObject().put(Ticket.STRUCTUREIDS, structureIds);
```
Comment (qpic-cgi) [inline src/main/java/net/atos/entng/support/services/impl/TicketServiceNeo4jImpl.java:95]: c'est pas une mauvaise chose de gérer le null pointer je pense

File: src/main/java/net/atos/entng/support/services/impl/TicketServiceSqlImpl.java [hunk 2]
```diff
@@ -265,9 +266,15 @@ public void listTickets(UserInfos user, Integer page, List<String> statuses, Lis
...
 
 
-		if(ALLOWED_SORT_BY_COLUMN.contains(sortBy)){
-			query.append(" ORDER BY t.");
-			query.append(sortBy);
+		if (ALLOWED_SORT_BY_COLUMN.contains(sortBy)) {
+			if (Objects.equals(sortBy, Ticket.SCHOOL_ID)) {
+				query.append(" ORDER BY array_position(" +
+						Sql.arrayPrepared(orderedStructures) + ",school_id)");
+				values.addAll(orderedStructures);
```
Comment (qpic-cgi) [inline src/main/java/net/atos/entng/support/services/impl/TicketServiceSqlImpl.java:273]: le null pointer peut être cool à gérer !

File: src/main/java/net/atos/entng/support/controllers/TicketController.java [hunk 3]
```diff
@@ -491,13 +491,20 @@ public void listTickets(final HttpServerRequest request) {
...
-                    // getting the profile for users
-                    ticketsPromise.future()
+                    Promise<JsonArray> promise = Promise.promise();
+                    promise.future()
+                            // getting the profile for users
                             .compose(tickets -> ticketService.getProfileFromTickets(tickets, i18nConfig))
                             .onSuccess(result -> renderJson(request, result))
                             .onFailure(err -> renderError(request, new JsonObject()));
+                    if (Objects.equals(sortBy, Ticket.SCHOOL_ID)) {
+                        ticketService.sortSchoolByName(user.getStructures())
```
[the remark this answers is not in the document]
Comment (Ozen40) [reply src/main/java/net/atos/entng/support/controllers/TicketController.java:501]: Cela va être fait dans le [CO-1796](https://jira.support-ent.fr/browse/CO-1796)

File: src/main/resources/public/ts/models/model.ts [hunk 4]
```diff
@@ -268,7 +268,12 @@ model.build = function() {
 				if(typeof callback === 'function'){
 					callback();
 				}
-			}.bind(this));
+			}.bind(this))
+				.error(function(err){
+				if (!!err && !!err.responseJSON && err.responseJSON.error && !!err.responseJSON.error.i18n){
```
Comment (Ozen40) [inline src/main/resources/public/ts/models/model.ts:273]: Exemple pris de : https://github.com/opendigitaleducation/entcore/blob/473bc7ea145cde961af7e3155744b56300b1f364/directory/src/main/resources/public/ts/model.ts#L727

Comment (qpic-cgi) [reply src/main/resources/public/ts/models/model.ts:273]:
```suggestion
				if (!!err && !!err.responseJSON && !!err.responseJSON.error && !!err.responseJSON.error.i18n){
```

File: src/main/resources/public/ts/models/model.ts [hunk 5]
```diff
@@ -268,7 +268,12 @@ model.build = function() {
...
+					notify.error('support.error.sort.tickets.by.school.name');
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/models/model.ts:274]:
```suggestion
					notify.error(err.responseJSON.error.i18n);
```
?

File: src/main/java/net/atos/entng/support/services/impl/TicketServiceImpl.java [hunk 6]
```diff
@@ -99,9 +99,11 @@ public Future<JsonObject> listStructureChildren(String schoolId) {
...
     public Future<JsonObject> sortSchoolByName(List<String> structureIds) {
         Promise<JsonObject> promise = Promise.promise();
-        TicketServiceNeo4jImpl.sortSchoolByName(structureIds)
-                .onSuccess(promise::complete)
-                .onFailure(promise::fail);
+        if (structureIds != null && !structureIds.isEmpty()){
+            TicketServiceNeo4jImpl.sortSchoolByName(structureIds)
+                    .onSuccess(promise::complete)
+                    .onFailure(promise::fail);
+        } else promise.fail("support.error.sort.school.by.name");
```
Comment (qpic-cgi) [inline src/main/java/net/atos/entng/support/services/impl/TicketServiceImpl.java:106]: t'aurais pu utiliser ton nouvel enum d'erreurs non ?

**Review by qpic-cgi — CHANGES_REQUESTED** [review]

**Review by qpic-cgi — APPROVED** [review]

**Review by BastienBenard — APPROVED** [review]

**Pull request merged**